test/integration/userFlow.test.js

```diff
@@ -0,0 +1,187 @@
+    expect.assertions(2);
+
+    graphql(graphqlSchema, createUserQuery, rootValue, contextNoToken)
+      .then(({ data, errors }) => {
+        expect(data.createUser).toMatchSnapshot('no user');
+        expect(errors).toMatchSnapshot('not logged in');
+        return;
+      })
+      .then(done)
+      .catch(done);
+  });
+
+  it('should return null and an auth error with an invalid token', done => {
+    expect.assertions(2);
+
+    graphql(graphqlSchema, createUserQuery, rootValue, invalidContext)
+      .then(({ data, errors }) => {
+        expect(data.createUser).toMatchSnapshot('no user');
+        expect(errors).toMatchSnapshot('no auth error');
+        return;
+      })
+      .then(done)
+      .catch(done);
+  });
+
+  it('should create a user by query', done => {
+    expect.assertions(2);
+
+    graphql(graphqlSchema, createUserQuery, rootValue, validContextCharlie)
+      .then(({ data, errors }) => {
+        expect(data.createUser).toMatchSnapshot('new user');
+        expect(errors).toMatchSnapshot('no errors');
+        return;
+      })
+      .then(done)
+      .catch(done);
+  });
+
+  it('should not create a second user for the same context', done => {
+    expect.assertions(4);
+
+    graphql(graphqlSchema, createUserQuery, rootValue, validContextCharlie)
+      .then(async function({ data, errors }) {
+        expect(data.createUser).toMatchSnapshot('new user');
+        expect(errors).toMatchSnapshot('no errors');
+        const usersFound = await UserModel.find({
+          email: 'charlie@thebear.me'
+        }).exec();
+        expect(usersFound).toHaveLength(1);
+        expect(
+          usersFound.map(obj => pick(obj, ['name', 'email', 'accountLinkId']))
+        ).toMatchSnapshot('no duplicate users');
+        return;
+      })
+      .then(done)
+      .catch(done);
+  });
+});
+
+describe('getUser', () => {
+  it('should return null and an auth error without a token', done => {
+    expect.assertions(2);
+
+    graphql(graphqlSchema, expectedUserQuery, rootValue, contextNoToken)
+      .then(({ data, errors }) => {
+        expect(data.getUser).toMatchSnapshot('no user');
+        expect(errors).toMatchSnapshot('not logged in');
+        return;
+      })
+      .then(done)
+      .catch(done);
+  });
+
+  it('should return null and an auth error with an invalid token', done => {
+    expect.assertions(2);
+
+    graphql(graphqlSchema, expectedUserQuery, rootValue, invalidContext)
+      .then(({ data, errors }) => {
+        expect(data.getUser).toMatchSnapshot('no user');
+        expect(errors).toMatchSnapshot('no auth error');
+        return;
+      })
+      .then(done)
+      .catch(done);
+  });
+
+  it('should return a user after one has been created', done => {
+    expect.assertions(2);
+
+    graphql(graphqlSchema, expectedUserQuery, rootValue, validContextCharlie)
+      .then(({ data, errors }) => {
+        expect(data).toMatchSnapshot('user found');
```

Comment:
I find this much harder to read than a simple test that creates an object, uses that to create a user and validates by compares the result with the input.

Can you explain to me what added value of snapshots in this case are? I see people online rave about snapshot, and I get that it's awesome for front end testing. But what does it bring us here but harder to read code?

If it sounds like I'm nagging, I'm sorry.  I'm genuinely interested, and am probably missing the point! 😄

Reply:
So, from my understanding, a snapshot is a text based representation of the thing being `expect()`'ed. It can be the output of a function, the state of a variable, anything!. `toMatchSnapshot()` effectively diffs the thing being `expect()`'ed against the stored snapshot to see if anything have changed. If things have changed, the test fails.

The most common use case is that instead of writing n * assertions, we take a dump of the thing being asserted instead and diff subsequent test output against it.

I guess there is a higher level of abstraction that comes with snapshots, you do not get a clear idea of the shape of the output unless you go and read the `.snap.js` files.

Reply:
Ah interesting, thanks for the explanation. The n assertions would add value.